Introduce `LocalDefId` to `HirId` lookup table

marmeladema opened the pull request:
cc #50928 

r? @ecstatic-morse

marmeladema commented:
If thats ok @ecstatic-morse , i'd like to run a perf run.

ecstatic-morse commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 20f20ac7a07b7db98b26c50cf0926cc7a2f818db with merge ba85afd575ddccb47ea677511836c480c50c2b61...
<!-- homu: {"type":"TryBuildStarted","head_sha":"20f20ac7a07b7db98b26c50cf0926cc7a2f818db","merge_sha":"ba85afd575ddccb47ea677511836c480c50c2b61"} -->

ecstatic-morse commented:
BTW, you committed an `.orig` file by accident. I think you need to wait for the perf run to finish before pushing again though.

ecstatic-morse commented:
@marmeladema This is basically just #63849 right? @ljedrz did we come to a conclusion about the approach in https://github.com/rust-lang/rust/pull/63849#issuecomment-537016618?

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=30099)
Build commit: ba85afd575ddccb47ea677511836c480c50c2b61 (`ba85afd575ddccb47ea677511836c480c50c2b61`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=30099"},"merge_sha":"ba85afd575ddccb47ea677511836c480c50c2b61"} -->

rust-timer commented:
Queued ba85afd575ddccb47ea677511836c480c50c2b61 with parent 46e85b4328fe18492894093c1092dfe509df4370, future [comparison URL](https://perf.rust-lang.org/compare.html?start=46e85b4328fe18492894093c1092dfe509df4370&end=ba85afd575ddccb47ea677511836c480c50c2b61).

rust-timer commented:
Finished benchmarking try commit ba85afd575ddccb47ea677511836c480c50c2b61, [comparison URL](https://perf.rust-lang.org/compare.html?start=46e85b4328fe18492894093c1092dfe509df4370&end=ba85afd575ddccb47ea677511836c480c50c2b61).

ljedrz commented:
@ecstatic-morse I wasn't sure how to proceed there; there was a follow-up PR, https://github.com/rust-lang/rust/pull/65837, but it was closed due to inactivity.

marmeladema commented:
Rebased and removed the mistakenly added `.orig` file.

About the approach here, I think the first commit is valuable no matter what because it just removes some useless uses of `NodeId` and final conversion in `<clone|into>_outputs` methods. Its just cleanup from https://github.com/rust-lang/rust/pull/72402

The 2 following commits are about removing all uses of `as_local_node_id` which are not really needed and helps reduce the APIs involving `NodeId`.

And the last one is about replacing the now unused `def_id_to_node_id` internal table by a `LocalDefId` to `HirId` table in order to achieve direct conversion without having to use an intermediary table involving `NodeId`s.

ecstatic-morse commented:
> I wasn't sure how to proceed there; there was a follow-up PR, #65837, but it was closed due to inactivity.

Ugh, that's a bummer.

So @nikomatsakis (who reviewed #63849) and (@eddyb who authored #65837). AFAICT with my somewhat limited understanding of the current `HirId` situation, this PR basically does the same thing as #63849. It replaces the current `DefId` -> `NodeId` -> `HirId` lookup chain with a direct mapping from `DefId` -> `HirId` to try to limit the use of `NodeId`s late in the compilation pipeline. As @eddyb noted in https://github.com/rust-lang/rust/pull/63849#issuecomment-537016618, this is somewhat silly, since a `HirId` is just a `LocalDefId` combined with an `ItemLocalId`, so we should be able to map to a `HirId` by using a "default" `ItemLocalId` (hopefully I'm not misinterpreting that too badly). #63849 was closed with the idea that @ljedrz would pursue this idea instead with mentorship from @eddyb.

@eddyb opened #65837 to get this process started, but it was never seriously reviewed, so things have sort of stalled out. I guess I see two options:

- Merge this PR, which contains a bit of cleanup along with the change from #63849, open an issue for removing the  `DefId` -> `HirId` map entirely.
- Ask @marmeladema to split out the "cleanup" parts, merge those, and then try to get #63849 going again, leaving the `DefId` -> `NodeId` -> `HirId` system in place.

marmeladema commented:
@ecstatic-morse i think as a start, i can split out the cleanup commits in a separate PR in order to focus in the import bits. Should the cleanup include commit 1 or 1,2,3? Including the 3 commits would reduce the API surface involving `NodeId`

marmeladema commented:
I've split out the first commit into its own PR and i'll do the same for the next 2 commits if needed.

nikomatsakis commented:
Splitting out the commits seems good -- I think the appropriate next course of action would be a MCP. IIRC, the older attempts pre-dated the system, and a big part of the problem (for me) was that I never understood what was being proposed, not really. =) Or at least I remember feeling confused. A summary like the one that @ecstatic-morse just gave is very helpful-- I would be in favor of trying to adopt the "correct" scheme, but I think it'd be worth actually describing what that is apart from its implementation.

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/72636.

bors commented:
:umbrella: The latest upstream changes (presumably #72716) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -611,7 +617,7 @@ struct UseError<'a> {
     candidates: Vec<ImportSuggestion>,
     /// The `NodeId` of the module to place the use-statements in.
-    node_id: NodeId,
+    def_id: DefId,
```
```suggestion
    def_id: LocalDefId,
```
`NodeId`s are always local, so they can always be replaced with local `DefId`s (if they can be replaced with `DefId`s at all).

marmeladema replied:
So i tried but i don't _think_ its possible because data comes from `ModuleData::normal_ancestor_id` which must be a `DefId` can it can come from another crate.

petrochenkov replied:
Hmm, the code previously used `unwrap()` to get a `NodeId` from `parent_scope.module.normal_ancestor_id`.
I agree that it's more robust to use a `DefId` here then given that `parent_scope` may point to other crates with macro 2.0 (I think).

`UsePlacementFinder` should still use a `LocalDefId`, it should just skip walking the crate entirely if the passed `target_module` is foreign.

marmeladema replied:
Ok i'll check if i can change `UsePlacementFinder` to use `LocalDefId`

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -253,21 +253,27 @@ impl<'a> From<&'a ast::PathSegment> for Segment {
-    target_module: NodeId,
+struct UsePlacementFinder<'d> {
+    definitions: &'d Definitions,
+    target_module: DefId,
```
```suggestion
    target_module: LocalDefId,
```

marmeladema replied:
Same as previous comment

petrochenkov replied:
See https://github.com/rust-lang/rust/pull/72552#discussion_r432669795.

petrochenkov commented on src/librustc_resolve/build_reduced_graph.rs:
```diff
@@ -1196,7 +1202,12 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
-                self.insert_unused_macro(ident, item.id, span);
+                self.insert_unused_macro(
+                    ident,
+                    self.r.definitions.local_def_id(item.id),
```
`let def_id = self.r.definitions.local_def_id(item.id);` is already evaluated above.
You can move it slightly higher and reuse in all the places inside `fn insert_unused_macro`.

marmeladema replied:
Fixed in https://github.com/rust-lang/rust/pull/72750

petrochenkov commented on src/librustc_hir/definitions.rs:
```diff
@@ -457,7 +443,7 @@ impl Definitions {
         let def_id = LocalDefId { local_def_index: self.table.allocate(key, def_path_hash) };
 
-        assert_eq!(self.def_id_to_node_id.push(node_id), def_id);
+        assert_eq!(self.def_id_to_hir_id.push(None), def_id);
```
Not sure why we need these pushes with `None` if we re-populate the whole `def_id_to_hir_id` table below anyway.

petrochenkov commented:
The change looks entirely reasonably to me and it passes CI so it doesn't trip the `HirId` validation, so I'm not sure why it would need an MCP.

Also, the first two commits could also be landed separately if the third one causes some questions.

petrochenkov commented:
Looks like we are not too far from being able to eliminate `NodeId`s entirely.

`NodeId`s are assigned in `InvocationCollector` and `DefId`s are assigned in `DefCollector`, which run immediately one after another during expansion.
So `NodeId`s as they exist now are only needed during a very short window, which perhaps can be eliminated eventually by merging passes or some other approach.

marmeladema commented:
For ease of review, i'll split out the first two commits in a separate pull request.

bors commented:
:umbrella: The latest upstream changes (presumably #72768) made this pull request unmergeable. Please resolve the merge conflicts.

marmeladema commented:
Rebased on master

petrochenkov commented:
@marmeladema 
https://github.com/rust-lang/rust/pull/73178 requires keeping the map "`LocalDefId` -> `NodeId`", but only until creation of `HirId`s.

The reason is that early lints have to work with `NodeId`s until `HirId`s are created.
`(Local)DefId`s are too coarse-grained for that task.

petrochenkov commented:
The ultimate solution to this problem is probably creating `HirId` earlier than AST -> HIR lowering, somewhere around creating `DefId`s.

Otherwise there's always some window in which you have no option except that to use `NodeId`s if you need a fine-grained id.

marmeladema commented:
I think keeping the map `LocalDefId` -> `NodeId` before the lowering is totally fine. I just think after hir lowering, all `NodeId`s should be gone for good. Ultimately i kind of think this map should belong to the resolver itself since its rely tied to the resolver job.

petrochenkov commented:
>Ultimately i kind of think this map should belong to the resolver itself since its rely tied to the resolver job. 

Yeah, it should be ok to move it from `Definitions` to `Resolver`.

marmeladema commented:
I've worked on another approach that might be better than this one actually. Let's close this for now.
